wordpress-mobile/WordPress-iOS
# Don't Store WordPress.com Passwords

### Description:
In this Pull Request, we enhance the ViewControllers outlined below, so that whenever a WordPress.com URL needs authentication, we send over the *Bearer Token* via the HTTP headers.

This enables us to stop storing the WordPress.com password in the keychain, and also allows us to prevent login errors due to Multifactor-Enabled accounts: the app might not know the latest MFA code, but the OAuth token is expected to be valid.

Plus, **WPComOAuthController** has been enhanced, so the Single Sign On mechanism also relies on the Bearer Token (scenario outlined in Testing Flow F).

Closes #1220

#### Enhanced viewControllers:
- **WPWebViewController**
- **PostPreviewViewController**
- **StatsWebViewController**

-- 

### Testing Scenarios:
Below, we'll detail several testing flows that have been considered while implementing this enhancement.

#### Flow A: Private Blog
1. Set a dotcom 'Testing Blog' as private
2. Open My Sites > 'Testing Blog'
3. Tap over 'View Site'

*Expected Results:*
- The site is visible, and WordPress.com doesn't ask for a password

<br/>

#### Flow B: Web Stats
1. Open My Sites > 'Testing Blog'
2. Tap over 'Stats'
3. Scroll down, and tap over 'View Web Version'

*Expected Results:*
- The stats are visible, and WordPress.com doesn't ask for a password

<br/>

#### Flow C: Post Preview
1. Open My Sites > 'Testing Blog'
2. Tap over Blog Posts
3. Tap over any post
4. Press the right top eye button

*Expected Results:*
- The post is visible, and WordPress.com doesn't ask for a password

<br/>

#### Flow D: Edit Site
1. Fresh Install
2. Log into a self hosted 'Blog A'
3. Install 'Disable XML-RPC' plugin
4. Open wp-admin in Safari, and update your dotcom password
5. Open 'My Sites' tab
6. Tap over 'Blog A' row
7. Tap 'Settings'
8. Update your password.

*Expected Results:*
- The Self Hosted blog is expected to return an error with code 405 (XMLRPC Disabled). 
- The app should push a WebView with your wp-admin already logged in.

<br/>

#### Flow E: Install jetpack
1. Open wp-admin in Safari, and turn off Jetpack in your self hosted blog.
2. Log into your self hosted site in WPiOS
3. Tap over 'My Sites' > 'Self Hosted Blog'
4. Tap over 'Stats'
5. Tap over 'Install Jetpack'

*Expected Results:*
- The app should push a WebView with your Self Hosted blog already logged in, and the **Install Jetpack View** onscreen.

<br/>

#### Flow F: Single Sign On
1. Install WPiOS and Sign Into a dotcom account.
2. Add a [new testing WordPress App in the Developer Portal](https://developer.wordpress.com/apps/)
3. Open this url in Safari.app: `wordpress-oauth-v2://authorize?client_id=[App ID]&redirect_url=[Callback URL]`

*Expected Results:*
- WPiOS should be onscreen, and the user should be asked whether if he wants to authorize or not the app. 
- No login or password should be required.
